File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.Queue;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Arrays;
import java.util.stream.Collectors;

class Solution {
    private static class Process{
        int index, priority;
        public Process(int index, int priority){
            this.index = index;
            this.priority = priority;
        }
    }
    
    public int solution(int[] priorities, int location) {
        int answer = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>((o1,o2) -> o2-o1);
        pq.addAll(Arrays.stream(priorities).boxed().collect(Collectors.toList()));
        
        Queue<Process> q = new LinkedList<>();
        
        for(int i=0;i<priorities.length;i++){
            q.offer(new Process(i, priorities[i]));
        }
        
        while(!q.isEmpty()){
            Process now = q.poll();
            if(now.priority == pq.peek()){
                pq.poll();
                answer++;
                if(now.index == location){
                    break;
                }
            }
            else{
                q.offer(now);

            }
        }
        return answer;
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.kt
import java.util.*

class Solution {
    fun solution(priorities: IntArray, location: Int): Int {
        var answer = 0
        val waiting: Queue<Process> = LinkedList()

        for (index in priorities.indices) {
            waiting.offer(Process(index, priorities[index]))
        }

        val pq: PriorityQueue<Int> = PriorityQueue(Comparator<Int> { a, b -> b - a })
        pq.addAll(priorities.toList())

        while (true) {
            val now = waiting.poll()
            if (pq.peek() == now.priority) {
                answer++
                pq.poll()
                if(now.index == location) break
            }else{
                waiting.offer(now)
            }
        }
        return answer
    }

    data class Process(val index: Int, val priority: Int)
}
